File: problems/1171_remove_zero_sum_consecutive_nodes_from_linked_list/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>
#include <queue>
#include <stack>
#include <cmath>
#include <numeric>
#include <set>
#include <map>
using namespace std;

class Solution {
public:
    ListNode* removeZeroSumSublists(ListNode* head) {
        ListNode dummy(0);
        dummy.next = head;
        unordered_map<int, ListNode*> seen;
        int prefixSum = 0;
        for (ListNode* node = &dummy; node; node = node->next) {
            prefixSum += node->val;
            seen[prefixSum] = node;
        }
        prefixSum = 0;
        for (ListNode* node = &dummy; node; node = node->next) {
            prefixSum += node->val;
            node->next = seen[prefixSum]->next;
        }
        return dummy.next;
    }
};

File: problems/1171_remove_zero_sum_consecutive_nodes_from_linked_list/solution.py
from shared.python.data_structures import ListNode


class Solution:
    def removeZeroSumSublists(self, head: ListNode) -> ListNode:
        dummy = ListNode(0)
        dummy.next = head
        prefix_sum = 0
        seen = {}
        node = dummy
        while node:
            prefix_sum += node.val
            seen[prefix_sum] = node
            node = node.next
        prefix_sum = 0
        node = dummy
        while node:
            prefix_sum += node.val
            node.next = seen[prefix_sum].next
            node = node.next
        return dummy.next
